test/test_latest_time_policy.cpp

```diff
@@ -0,0 +1,237 @@
+using namespace message_filters::sync_policies;
+
+struct Header
+{
+  rclcpp::Time stamp;
+};
+
+struct Msg
+{
+  Header header;
+  int data;
+};
+typedef std::shared_ptr<Msg> MsgPtr;
+typedef std::shared_ptr<Msg const> MsgConstPtr;
+
+namespace message_filters
+{
+namespace message_traits
+{
+template<>
+struct TimeStamp<Msg>
+{
+  static rclcpp::Time value(const Msg& m)
+  {
+    return m.header.stamp;
+  }
+};
+}
+}
+
+class Helper
+{
+public:
+  Helper()
+  : count_(0)
+  {}
+
+  void cb(const MsgConstPtr& p, const MsgConstPtr& q, const MsgConstPtr& r)
+  {
+    ASSERT_TRUE(p);
+    ASSERT_TRUE(q);
+    ASSERT_TRUE(r);
+
+    p_ = p; q_ = q; r_ = r;
+    ++count_;
+  }
+
+  MsgConstPtr p_, q_, r_;
+  uint16_t count_;
+};
+
+typedef LatestTime<Msg, Msg, Msg> Policy3;
+typedef Synchronizer<Policy3> Sync3;
+
+TEST(LatestTime, Leading)
+{
+  Sync3 sync(1);
+  Helper h;
+  sync.registerCallback(std::bind(&Helper::cb, &h, std::placeholders::_1,
+                                                   std::placeholders::_2,
+                                                   std::placeholders::_3));
+  std::vector<MsgPtr> p; p.reserve(8U);
+  std::vector<MsgPtr> q; q.reserve(4U);
+  std::vector<MsgPtr> r; r.reserve(2U);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    MsgPtr p_idx(std::make_shared<Msg>()); p_idx->data = idx; p.push_back(p_idx);
+    if(idx % 2U == 0U)
+    {
+      MsgPtr q_idx(std::make_shared<Msg>()); q_idx->data = idx; q.push_back(q_idx);
+    }
+    if(idx % 4U == 0U)
+    {
+      MsgPtr r_idx(std::make_shared<Msg>()); r_idx->data = idx; r.push_back(r_idx);
+    }
+  }
+
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    if(idx % 2U == 0U)
+    {
+      sync.add<1>(q[idx/2]);
+    }
+    if(idx % 4U == 0U)
+    {
+      sync.add<2>(r[idx/4]);
+    }
+    sync.add<0>(p[idx]);
+    
+    ASSERT_EQ(h.count_, idx);
+    if(idx > 0)
+    {
+      ASSERT_EQ(h.p_->data, p[idx]->data);
+      ASSERT_EQ(h.q_->data, q[idx/2]->data);
+      ASSERT_EQ(h.r_->data, r[idx/4]->data);
+    }
+    else
+    {
+      ASSERT_FALSE(h.p_);
+      ASSERT_FALSE(h.q_);
+      ASSERT_FALSE(h.r_);
+    }
+
+    rate.sleep();
+  }
+}
+
+TEST(LatestTime, Trailing)
+{
+  Sync3 sync(1);
+  Helper h;
+  sync.registerCallback(std::bind(&Helper::cb, &h, std::placeholders::_1,
+                                                   std::placeholders::_2,
+                                                   std::placeholders::_3));
+  std::vector<MsgPtr> p; p.reserve(8U);
+  std::vector<MsgPtr> q; q.reserve(4U);
+  std::vector<MsgPtr> r; r.reserve(2U);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    MsgPtr p_idx(std::make_shared<Msg>()); p_idx->data = idx; p.push_back(p_idx);
+    if(idx % 2U == 0U)
+    {
+      MsgPtr q_idx(std::make_shared<Msg>()); q_idx->data = idx; q.push_back(q_idx);
+    }
+    if(idx % 4U == 0U)
+    {
+      MsgPtr r_idx(std::make_shared<Msg>()); r_idx->data = idx; r.push_back(r_idx);
+    }
+  }
```

Comment:
Both tests are the same up to here. Consider moving the common code to a function. Maybe even to `LatestTime::SetUp` so it's called at the beginning of every test.

Reply:
I pushed it to a Test Fixture SetUp.